internal/process/process.go

```diff
@@ -100,18 +100,20 @@ func (p *Process) cmdStart() error {
 
 // cmdLoop watches an already running process, restarting it when appropriate.
 func (p *Process) cmdLoop(ctx context.Context) error {
+	processShutdownCh := make(chan struct{})
```

Comment:
this fixes a race condition around checking p.Cmd from a goroutine

Reply:
there's already a context that conveys the intent on whether or not the process was asked to shut down, a second context seems confusing at the least, and the semantics don't line up (nothing is being canceled?). Perhaps `close` isn't a whole lot better, though. Reusing the existing context here would be incorrect, as it represents the request and intent to shut down. Here I want to know "has the external process I'm running stopped yet?".

You're probably right that it's still possible that the process closes right before the `.Kill` is called. It seems like this would be inherently racey anyway. Not sure how you would resolve that. By definition the goroutine that calls `cmd.Process.Kill()` can't be the same one that called `cmd.Wait()`.

We might not need the nil check on `p.Stdin` anymore. It used to be that you could stop the process before it finished starting up, but that code is in New now, and it never gets unset. There's no race there, but we can probably remove the check, now. Even if cmdStart is called again, we don't (and don't need to) reassign this. 

You can't move the `p.Stdin.Close()` and `gracefulStop()` into the same goroutine as the cmdLoop (because that's what triggers the very start of the shutdown process; p.cmdWait will never return without this), but it looks like you could move `p.Stdin.Close()` to the Stop() function. This is probably more clear that it is involved at the start of triggering a normal stopping process. I'll try that.